Update dependencies and use more specific versions for dependencies with known CVEs

jsdanielh opened the pull request:
## What's in this pull request?

- Use a more specific version for the `rand_core` dependency than versions with known CVEs.
  This fixes #2454.
- Use a more specific version for the `futures-util` dependency than versions with known CVEs.
  This fixes #2457.
- Use a more specific version for the `tracing` dependency than versions with known CVEs.
  This fixes #2453.
- Use a more specific version for the `regex` dependency than versions with known CVEs.
- Update the `itertools` dependency to the latest version (0.12.1) as for some reason this hasn't been flagged by GH dependabot.
- Update the `hyper` dependency to the latest version (1.3.1) for the `metrics-server` subcrate.
- Updated the `crossbeam-channel` dependency from version 0.5.7 to 0.5.12 (latest) since the former version was already yanked as pointed by `cargo audit`.
- Switch to a maintained version of the `dotenv` dependency: `dotenvy` since the former is flagged by `cargo audit` as unmaintained.
- Switch to a maintained version of the `ansi-term` dependency: `ansiterm` since the former was flagged by `cargo audit` as unmaintained.
- Add Github workflow to check for security vulnerabilities using `cargo audit`.
  This workflow was set up to run every Wednesday at 2:00 UTC.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
keys: Use a more specific version for `rand_core`

Use a more specific version for the `rand_core` dependency than
versions with known CVEs.
This fixes #2454.

jsdanielh committed:
Use a more specific version for `futures-util`

Use a more specific version for the `futures-util` dependency than
versions with known CVEs.
This fixes #2457.

jsdanielh committed:
Use a more specific version for `tracing`

Use a more specific version for the `tracing` dependency than
versions with known CVEs.
This fixes #2453.

jsdanielh committed:
key-derivation: Use a more specific version for `regex`

Use a more specific version for the `regex` dependency than versions
with known CVEs.

jsdanielh committed:
keys: Update the `itertools` depdency to the latest version

Update the `itertools` dependency to the latest version (0.12.1) as
for some reason this hasn't been flagged by GH dependabot.

jsdanielh committed:
metrics-server: Update the `hyper` depdency to the latest version

Update the `hyper` dependency to the latest version (1.3.1) for the
`metrics-server` subcrate.

hrxi commented on keys/Cargo.toml:
```diff
@@ -30,14 +30,14 @@ curve25519-dalek = { version = "4", features = [
 p256 = "0.13"
 rand = "0.8"
-rand_core = "0.6"
+rand_core = "0.6.4"
```
I don't think this is a worthwhile thing to get into. As I've already mentioned privately, we publish a lock file, so a outdated lock file for a contributor is an unlikely problem. What remains is that we're suppose to update the version numbers in `Cargo.toml` each time there's a CVE in any of our direct dependencies. With no real upside, this sounds like unnecessary work we're committing to.

If we just do this because we really want to tick a check box, then … I guess we can do it.

hrxi replied:
We talked about this in a meeting and @jsdanielh reminded me that this might be useful for Rust projects depending on our libraries directly. I missed that and then the change makes sense.

hrxi commented:
Could we perhaps move these dependencies to the global `Cargo.toml` so that we don't have to update the version number in several places?

How do we notice that we specify a version in `Cargo.toml` that is lower than a CVE?

Eligioo commented:
@hrxi  [deps.rs](https://deps.rs/repo/github/nimiq/core-rs-albatross#vulnerabilities) is one effort I know though it doesn't actively notify you about versions lower than a CVE. But it's a badge in the README of this repo already.

`futures-util` doesn't show up there for example because it's already locked at 0.3.30.

https://github.com/rustsec/audit-check might be interesting

jsdanielh commented:
> Could we perhaps move these dependencies to the global `Cargo.toml` so that we don't have to update the version number in several places?

Done. Only did it for `futures` and `log` since the rest only seemed to be use in a single crate.
 
> How do we notice that we specify a version in `Cargo.toml` that is lower than a CVE?

Yeah, as @Eligioo mentioned we already have [deps.rs](https://deps.rs/repo/github/nimiq/core-rs-albatross#vulnerabilities) and then there is also `cargo audit` which I can see if I can add it to GH actions.

jsdanielh committed:
Update the `crossbeam-channel` dependency to the latest version

Updated the `crossbeam-channel` dependency from version 0.5.7 to
0.5.12 (latest) since the former version was already yanked as
pointed by `cargo audit`.

jsdanielh committed:
rpc-client: Switch to a maintained version of the `dotenv` dependency

Switch to a maintained version of the `dotenv` dependency: `dotenvy`
since the former is flagged by `cargo audit` as unmaintained.

jsdanielh committed:
log: Switch to a maintained version of the `ansi-term` dependency

Switch to a maintained version of the `ansi-term` dependency:
`ansiterm` since the former was flagged by `cargo audit` as
unmaintained.

jsdanielh committed:
Add GH workflow to check for security vulnerabilities

Add Github workflow to check for security vulnerabilities using
`cargo audit`.
This workflow was set up to run every Wednesday at 2:00 UTC.

jsdanielh commented:
> > Could we perhaps move these dependencies to the global `Cargo.toml` so that we don't have to update the version number in several places?
> 
> Done. Only did it for `futures` and `log` since the rest only seemed to be use in a single crate.
> 
> > How do we notice that we specify a version in `Cargo.toml` that is lower than a CVE?
> 
> Yeah, as @Eligioo mentioned we already have [deps.rs](https://deps.rs/repo/github/nimiq/core-rs-albatross#vulnerabilities) and then there is also `cargo audit` which I can see if I can add it to GH actions.

Done. Just added a GH action that is going to be run periodically to check for `cargo audit` warnings.

Eligioo commented on .github/workflows/security_audit.yml:
```diff
@@ -0,0 +1,14 @@
+  audit:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v3
```
Can you set `actions/checkout` to v4?

hrxi commented on web-client/Cargo.toml:
```diff
@@ -23,10 +23,10 @@ crate-type = ["cdylib"]
 [dependencies]
 futures = "0.3"
-futures-util = "0.3"
+futures-util = "0.3.30"
```
```suggestion
futures = { workspace = true }
```

hrxi reviewed (approved):
LGTM otherwise.
